File: hydra-haskell/src/main/haskell/Hydra/Dsl/Meta/Lib/Chars.hs
-- | Phantom-typed term DSL for the hydra.lib.chars library

module Hydra.Dsl.Meta.Lib.Chars where

import Hydra.Phantoms
import Hydra.Dsl.Meta.Phantoms
import qualified Hydra.Dsl.Terms as Terms
import Hydra.Sources.Libraries


-- | Check whether a character is alphanumeric.
isAlphaNum :: TTerm Int -> TTerm Bool
isAlphaNum = primitive1 _chars_isAlphaNum

-- | Check whether a character is lowercase.
isLower :: TTerm Int -> TTerm Bool
isLower = primitive1 _chars_isLower

-- | Check whether a character is a whitespace character.
isSpace :: TTerm Int -> TTerm Bool
isSpace = primitive1 _chars_isSpace

-- | Check whether a character is uppercase.
isUpper :: TTerm Int -> TTerm Bool
isUpper = primitive1 _chars_isUpper

-- | Convert a character to lowercase.
toLower :: TTerm Int -> TTerm Int
toLower = primitive1 _chars_toLower

-- | Convert a character to uppercase.
toUpper :: TTerm Int -> TTerm Int
toUpper = primitive1 _chars_toUpper

File: packages/hydra-java/src/main/java/hydra/dsl/meta/lib/Chars.java
package hydra.dsl.meta.lib;
import hydra.dsl.meta.Phantoms;
import hydra.typed.TypedTerm;

/** Phantom-typed term DSL for the {@code hydra.lib.chars} library. */
public final class Chars {
    private Chars() {}

    private static <A> TypedTerm<A> prim(String name) {
        return Phantoms.var("hydra.lib.chars." + name);
    }

    public static <A> TypedTerm<A> isAlphaNum(TypedTerm<?> c) { return Phantoms.apply(prim("isAlphaNum"), c); }
    public static <A> TypedTerm<A> isLower(TypedTerm<?> c) { return Phantoms.apply(prim("isLower"), c); }
    public static <A> TypedTerm<A> isSpace(TypedTerm<?> c) { return Phantoms.apply(prim("isSpace"), c); }
    public static <A> TypedTerm<A> isUpper(TypedTerm<?> c) { return Phantoms.apply(prim("isUpper"), c); }
    public static <A> TypedTerm<A> toLower(TypedTerm<?> c) { return Phantoms.apply(prim("toLower"), c); }
    public static <A> TypedTerm<A> toUpper(TypedTerm<?> c) { return Phantoms.apply(prim("toUpper"), c); }
}
